In [ ]:
# pip install pandas
# this will install pandas library in Python.
# But you dont need it

In [ ]:
# import library
import pandas

In [ ]:
pandas.__version__

'2.2.2'

In [ ]:
data = {
  "calories" : [420, 380, 390],
  "duration" : [50, 40, 45]
}
# This is a dictionary with list

In [ ]:
import pandas as pd
pd.DataFrame(data)
# create a dataframe from a dict

calories  duration
0       420        50
1       380        40
2       390        45

In [ ]:
from google.colab import files
uploaded = files.upload()

Saving student.csv to student (1).csv


In [ ]:
df = pd.read_csv("student.csv")

In [ ]:
df = pd.read_excel("student.xlsx")

In [ ]:
df.head()
# return first 5 rows


id        name  class  mark  gender
0   1    John Deo   Four    75  female
1   2    Max Ruin  Three    85    male
2   3      Arnold  Three    55    male
3   4  Krish Star   Four    60  female
4   5   John Mike   Four    60  female

In [ ]:
df.head(3)
# first 3 rows


id      name  class  mark  gender
0   1  John Deo   Four    75  female
1   2  Max Ruin  Three    85    male
2   3    Arnold  Three    55    male

In [ ]:
df.tail(1)
# last row


id        name class  mark  gender
34  35  Rows Noump   Six    88  female

In [ ]:
df.sample()

id    name  class  mark gender
2   3  Arnold  Three    55   male

In [ ]:
df.sample(4)

id         name  class  mark  gender
30  31  Marry Toeey   Four    88    male
6    7  My John Rob  Fifth    78    male
5    6    Alex John   Four    55    male
0    1     John Deo   Four    75  female

In [ ]:
# how many rows and columns are there?
df.shape
# there are 35 rows and 5 column

(35, 5)

In [ ]:
df.columns
#name of the columns


Index(['id', 'name', 'class', 'mark', 'gender'], dtype='object')

In [ ]:
df.describe()
# info about numeric columns

id       mark
count  35.000000  35.000000
mean   18.000000  74.657143
std    10.246951  16.401117
min     1.000000  18.000000
25%     9.500000  62.500000
50%    18.000000  79.000000
75%    26.500000  88.000000
max    35.000000  96.000000



*   std = standart deviation

*   25% of the student has mark 62.5 or less
*   50% of the students has mark 79 or less


*   this is useful to identify outlier






In [ ]:
df.dtypes

id         int64
name      object
class     object
mark       int64
gender    object
dtype: object

In [ ]:
df.isnull().sum()
# number of missing value per column

id        0
name      1
class     1
mark      0
gender    2
dtype: int64

In [ ]:
# return names only
df["name"]

0        John Deo
1        Max Ruin
2          Arnold
3      Krish Star
4       John Mike
5       Alex John
6     My John Rob
7          Asruid
8         Tes Qry
9        Big John
10         Ronald
11          Recky
12            Kty
13           Bigy
14       Tade Row
15          Gimmy
16          Tumyu
17          Honny
18          Tinny
19         Jackly
20     Babby John
21         Reggid
22          Herod
23      Tiddy Now
24       Giff Tow
25         Crelea
26            NaN
27      Rojj Base
28    Tess Played
29      Reppy Red
30    Marry Toeey
31      Binn Rott
32      Kenn Rein
33       Gain Toe
34     Rows Noump
Name: name, dtype: object

In [ ]:
df.mark
#this is dot notation. but it may not work if column name is sames as one of keyword in Python.

0     75
1     85
2     55
3     60
4     60
5     55
6     78
7     85
8     78
9     55
10    89
11    94
12    88
13    88
14    88
15    88
16    54
17    75
18    18
19    65
20    69
21    55
22    79
23    78
24    88
25    79
26    81
27    86
28    55
29    79
30    88
31    90
32    96
33    69
34    88
Name: mark, dtype: int64

In [ ]:
df.class
#because class is reversed keyword

SyntaxError: invalid syntax (<ipython-input-24-350afe6b7ecd>, line 1)

In [ ]:
rows, columns = df.shape

print(f"Rows: {rows}")
print(f"Columns: {columns}")

Rows: 35
Columns: 5


In [ ]:
df["class"].head()

0     Four
1    Three
2    Three
3     Four
4     Four
Name: class, dtype: object

In [ ]:
df["mark"].mean()
#calculate average mark

np.float64(74.65714285714286)

In [ ]:
df["mark"].head().median()
# median mark for first 5 student in this class

60.0

In [ ]:
df[["class", "name"]].head()
# you can write only 1 column in slicing method
# if you need to return more than 1 column, include them in a list
# first [] used for slicing
# second[] used for list of the column names

class        name
0   Four    John Deo
1  Three    Max Ruin
2  Three      Arnold
3   Four  Krish Star
4   Four   John Mike

Write syntax to return Name, gender and mark of 3 random students.

In [ ]:
df[["name", "gender", "mark"]].sample(3)

name  gender  mark
29  Reppy Red  female    79
33   Gain Toe    male    69
23  Tiddy Now    male    78

In [ ]:
df.loc[0:4, "name"]
# [rows to return , column names]

0      John Deo
1      Max Ruin
2        Arnold
3    Krish Star
4     John Mike
Name: name, dtype: object

In [ ]:
df.iloc[0:5, 1]
# 1 is the index name of the column

0      John Deo
1      Max Ruin
2        Arnold
3    Krish Star
4     John Mike
Name: name, dtype: object

Recap Exercises
1) Return name and class of first 3 students


Use slicing method
Use loc
Use iloc


In [ ]:
df.loc[0:2, ["name", "class"]]

name  class
0  John Deo   Four
1  Max Ruin  Three
2    Arnold  Three

2) Return average mark for all students

In [ ]:
df["mark"].mean()

np.float64(74.65714285714286)

3) Return 3 sample class and gender of the students

In [ ]:
df[["class","gender"]].sample(3)

class  gender
34    Six  female
28  Seven    male
9    Four  female

4) Return first 5 rows

In [ ]:
df.head()

id        name  class  mark  gender
0   1    John Deo   Four    75  female
1   2    Max Ruin  Three    85    male
2   3      Arnold  Three    55    male
3   4  Krish Star   Four    60  female
4   5   John Mike   Four    60  female

first column index 0


same for first row index 0

In [ ]:
# Return male student?
df[df["gender"]== "male"]

id         name  class  mark gender
1    2     Max Ruin  Three    85   male
2    3       Arnold  Three    55   male
5    6    Alex John   Four    55   male
6    7  My John Rob  Fifth    78   male
7    8       Asruid   Five    85   male
14  15     Tade Row    NaN    88   male
15  16        Gimmy   Four    88   male
16  17        Tumyu    Six    54   male
17  18        Honny   Five    75   male
18  19        Tinny   Nine    18   male
22  23        Herod  Eight    79   male
23  24    Tiddy Now  Seven    78   male
24  25     Giff Tow  Seven    88   male
25  26       Crelea  Seven    79   male
28  29  Tess Played  Seven    55   male
30  31  Marry Toeey   Four    88   male
33  34     Gain Toe  Seven    69   male

* Write code to return student class of six?

In [ ]:
df[df["class"] == "Six"]

id        name class  mark  gender
8    9     Tes Qry   Six    78     NaN
10  11      Ronald   Six    89  female
11  12       Recky   Six    94  female
16  17       Tumyu   Six    54    male
29  30   Reppy Red   Six    79  female
32  33   Kenn Rein   Six    96  female
34  35  Rows Noump   Six    88  female

* Return student with mark less than 52.

In [ ]:
df[df["mark"] < 52]

id   name class  mark gender
18  19  Tinny  Nine    18   male

In [ ]:
df.loc[df["mark"]<52]

id   name class  mark gender
18  19  Tinny  Nine    18   male

- Return female students with their mark greater than 80


In [ ]:
df[(df["gender"]=="female") & (df["mark"]>80)]

id        name  class  mark  gender
10  11      Ronald    Six    89  female
11  12       Recky    Six    94  female
12  13         Kty  Seven    88  female
13  14        Bigy  Seven    88  female
27  28   Rojj Base  Seven    86  female
31  32   Binn Rott  Seven    90  female
32  33   Kenn Rein    Six    96  female
34  35  Rows Noump    Six    88  female

In [ ]:
df.loc[ (df["gender"]=="female") & (df["mark"]>80) ]

id        name  class  mark  gender
10  11      Ronald    Six    89  female
11  12       Recky    Six    94  female
12  13         Kty  Seven    88  female
13  14        Bigy  Seven    88  female
27  28   Rojj Base  Seven    86  female
31  32   Binn Rott  Seven    90  female
32  33   Kenn Rein    Six    96  female
34  35  Rows Noump    Six    88  female

In [ ]:
df.loc[ (df["gender"]=="female") & (df["mark"]>80) ].head()

id       name  class  mark  gender
10  11     Ronald    Six    89  female
11  12      Recky    Six    94  female
12  13        Kty  Seven    88  female
13  14       Bigy  Seven    88  female
27  28  Rojj Base  Seven    86  female

In [ ]:
# female, mark > 80 and Seven
df.loc[ (df["gender"]=="female") & (df["mark"]>80) & (df["class"]=="Seven") , ["name", "gender"] ]

name  gender
12        Kty  female
13       Bigy  female
27  Rojj Base  female
31  Binn Rott  female

In [ ]:
# female, mark > 80 and Seven
df.loc[ (df["gender"]=="female") & (df["mark"]>80) & (df["class"]=="Seven")][["name", "gender"]]

name  gender
12        Kty  female
13       Bigy  female
27  Rojj Base  female
31  Binn Rott  female

In [ ]:
df[(df["gender"] == "male") | (df["class"]=="Five")]

id         name  class  mark gender
1    2     Max Ruin  Three    85   male
2    3       Arnold  Three    55   male
5    6    Alex John   Four    55   male
6    7  My John Rob  Fifth    78   male
7    8       Asruid   Five    85   male
14  15     Tade Row    NaN    88   male
15  16        Gimmy   Four    88   male
16  17        Tumyu    Six    54   male
17  18        Honny   Five    75   male
18  19        Tinny   Nine    18   male
22  23        Herod  Eight    79   male
23  24    Tiddy Now  Seven    78   male
24  25     Giff Tow  Seven    88   male
25  26       Crelea  Seven    79   male
28  29  Tess Played  Seven    55   male
30  31  Marry Toeey   Four    88   male
33  34     Gain Toe  Seven    69   male

Write Python code to calculate



Average mark across all students


In [ ]:
df["mark"].mean()

np.float64(74.65714285714286)

Average mark across male and female students

In [ ]:
df.groupby("gender")["mark"].mean()

gender
female    77.312500
male      71.588235
Name: mark, dtype: float64

Max and min mark

In [ ]:
df["mark"].max() , df["mark"].min()

(96, 18)

Max mark for male and female students

In [ ]:
df.groupby("gender")["mark"].max()

gender
female    96
male      88
Name: mark, dtype: int64

Average mark for class 'Five'

In [ ]:
df[df["class"]=="Five"]["mark"].mean()

np.float64(80.0)

Average mark for class 'Six' and male students only

In [ ]:
df.loc[ (df["class"]=="Six") & (df["gender"]=="male")]["mark"].mean()

np.float64(54.0)

In [ ]:
df.loc[df["name"].str.contains("John", na=False)]
#.str will convert Pandas series to string object
# na=False means skip missing values

id         name  class  mark  gender
0    1     John Deo   Four    75  female
4    5    John Mike   Four    60  female
5    6    Alex John   Four    55    male
6    7  My John Rob  Fifth    78    male
9   10     Big John   Four    55  female
20  21   Babby John   Four    69  female

In [ ]:
df.sort_values(by="mark")
# sort results by mark column

id         name  class  mark  gender
18  19        Tinny   Nine    18    male
16  17        Tumyu    Six    54    male
2    3       Arnold  Three    55    male
5    6    Alex John   Four    55    male
28  29  Tess Played  Seven    55    male
21  22       Reggid  Seven    55  female
9   10     Big John   Four    55  female
4    5    John Mike   Four    60  female
3    4   Krish Star   Four    60  female
19  20       Jackly   Nine    65  female
20  21   Babby John   Four    69  female
33  34     Gain Toe  Seven    69    male
0    1     John Deo   Four    75  female
17  18        Honny   Five    75    male
8    9      Tes Qry    Six    78     NaN
23  24    Tiddy Now  Seven    78    male
6    7  My John Rob  Fifth    78    male
22  23        Herod  Eight    79    male
29  30    Reppy Red    Six    79  female
25  26       Crelea  Seven    79    male
26  27          NaN  Three    81     NaN
1    2     Max Ruin  Three    85    male
7    8       Asruid   Five    85    male
27  28    Rojj Base  Seven    86  female
12  13          Kty  Seven    88  female
24  25     Giff Tow  Seven    88    male
15  16        Gimmy   Four    88    male
14  15     Tade Row    NaN    88    male
30  31  Marry Toeey   Four    88    male
13  14         Bigy  Seven    88  female
34  35   Rows Noump    Six    88  female
10  11       Ronald    Six    89  female
31  32    Binn Rott  Seven    90  female
11  12        Recky    Six    94  female
32  33    Kenn Rein    Six    96  female

In [ ]:
df.loc[:, "name"].sort_values()
# you can ignore by argument if there is only 1 column in the output

5       Alex John
2          Arnold
7          Asruid
20     Babby John
9        Big John
13           Bigy
31      Binn Rott
25         Crelea
33       Gain Toe
24       Giff Tow
15          Gimmy
22          Herod
17          Honny
19         Jackly
0        John Deo
4       John Mike
32      Kenn Rein
3      Krish Star
12            Kty
30    Marry Toeey
1        Max Ruin
6     My John Rob
11          Recky
21         Reggid
29      Reppy Red
27      Rojj Base
10         Ronald
34     Rows Noump
14       Tade Row
8         Tes Qry
28    Tess Played
23      Tiddy Now
18          Tinny
16          Tumyu
26            NaN
Name: name, dtype: object

In [ ]:
df.loc[1:10, ["name", "class"]].sort_values(by="class")

name  class
6   My John Rob  Fifth
7        Asruid   Five
3    Krish Star   Four
4     John Mike   Four
5     Alex John   Four
9      Big John   Four
8       Tes Qry    Six
10       Ronald    Six
1      Max Ruin  Three
2        Arnold  Three

In [ ]:
df.sort_values(by="mark", ascending=False)

id         name  class  mark  gender
32  33    Kenn Rein    Six    96  female
11  12        Recky    Six    94  female
31  32    Binn Rott  Seven    90  female
10  11       Ronald    Six    89  female
30  31  Marry Toeey   Four    88    male
34  35   Rows Noump    Six    88  female
24  25     Giff Tow  Seven    88    male
14  15     Tade Row    NaN    88    male
15  16        Gimmy   Four    88    male
12  13          Kty  Seven    88  female
13  14         Bigy  Seven    88  female
27  28    Rojj Base  Seven    86  female
7    8       Asruid   Five    85    male
1    2     Max Ruin  Three    85    male
26  27          NaN  Three    81     NaN
29  30    Reppy Red    Six    79  female
25  26       Crelea  Seven    79    male
22  23        Herod  Eight    79    male
6    7  My John Rob  Fifth    78    male
23  24    Tiddy Now  Seven    78    male
8    9      Tes Qry    Six    78     NaN
17  18        Honny   Five    75    male
0    1     John Deo   Four    75  female
33  34     Gain Toe  Seven    69    male
20  21   Babby John   Four    69  female
19  20       Jackly   Nine    65  female
3    4   Krish Star   Four    60  female
4    5    John Mike   Four    60  female
2    3       Arnold  Three    55    male
5    6    Alex John   Four    55    male
9   10     Big John   Four    55  female
21  22       Reggid  Seven    55  female
28  29  Tess Played  Seven    55    male
16  17        Tumyu    Six    54    male
18  19        Tinny   Nine    18    male

In [ ]:
df.sort_values(by="mark", ascending=False).head()

id         name  class  mark  gender
32  33    Kenn Rein    Six    96  female
11  12        Recky    Six    94  female
31  32    Binn Rott  Seven    90  female
10  11       Ronald    Six    89  female
30  31  Marry Toeey   Four    88    male

In [ ]:
df.loc[:, :].sort_values(by="gender", ascending=False)

id         name  class  mark  gender
17  18        Honny   Five    75    male
22  23        Herod  Eight    79    male
24  25     Giff Tow  Seven    88    male
25  26       Crelea  Seven    79    male
18  19        Tinny   Nine    18    male
1    2     Max Ruin  Three    85    male
16  17        Tumyu    Six    54    male
15  16        Gimmy   Four    88    male
14  15     Tade Row    NaN    88    male
28  29  Tess Played  Seven    55    male
23  24    Tiddy Now  Seven    78    male
30  31  Marry Toeey   Four    88    male
7    8       Asruid   Five    85    male
6    7  My John Rob  Fifth    78    male
5    6    Alex John   Four    55    male
33  34     Gain Toe  Seven    69    male
2    3       Arnold  Three    55    male
27  28    Rojj Base  Seven    86  female
31  32    Binn Rott  Seven    90  female
32  33    Kenn Rein    Six    96  female
29  30    Reppy Red    Six    79  female
0    1     John Deo   Four    75  female
21  22       Reggid  Seven    55  female
20  21   Babby John   Four    69  female
19  20       Jackly   Nine    65  female
13  14         Bigy  Seven    88  female
12  13          Kty  Seven    88  female
11  12        Recky    Six    94  female
10  11       Ronald    Six    89  female
9   10     Big John   Four    55  female
4    5    John Mike   Four    60  female
3    4   Krish Star   Four    60  female
34  35   Rows Noump    Six    88  female
8    9      Tes Qry    Six    78     NaN
26  27          NaN  Three    81     NaN

In [ ]:
## How many male and female students are there?
df["gender"].value_counts()
# how many times male is in the gender

gender
male      17
female    16
Name: count, dtype: int64

In [ ]:
# number of student in each class
df["class"].value_counts()

class
Seven    10
Four      8
Six       7
Three     3
Nine      2
Five      2
Fifth     1
Eight     1
Name: count, dtype: int64